src/useTimeout.ts

```diff
@@ -47,33 +48,60 @@ function setChainedTimeout(
  * ```
  */
 export default function useTimeout() {
+  const [timeout, setTimeoutState] = useState<TimeoutState | null>(null)
   const isMounted = useMounted()
 
   // types are confused between node and web here IDK
-  const handleRef = useRef<any>()
+  const handleRef = useRef<any>(null)
 
-  useWillUnmount(() => clearTimeout(handleRef.current))
+  useEffect(() => {
+    if (!timeout) {
+      return
+    }
 
-  return useMemo(() => {
-    const clear = () => clearTimeout(handleRef.current)
+    const { fn, delayMs } = timeout
 
-    function set(fn: () => void, delayMs = 0): void {
-      if (!isMounted()) return
+    function task() {
+      if (isMounted()) {
+        setTimeoutState(null)
+      }
+      fn()
+    }
 
-      clear()
+    if (delayMs <= MAX_DELAY_MS) {
+      // For simplicity, if the timeout is short, just set a normal timeout.
+      handleRef.current = setTimeout(task, delayMs)
+    } else {
+      setChainedTimeout(handleRef, task, Date.now() + delayMs)
+    }
+    const handle = handleRef.current
 
-      if (delayMs <= MAX_DELAY_MS) {
-        // For simplicity, if the timeout is short, just set a normal timeout.
-        handleRef.current = setTimeout(fn, delayMs)
-      } else {
-        setChainedTimeout(handleRef, fn, Date.now() + delayMs)
+    return () => {
+      // this should be a no-op since they are either the same or `handle`
+      // already expired but no harm in calling twice
+      if (handleRef.current !== handle) {
+        clearTimeout(handle)
       }
+
+      clearTimeout(handleRef.current)
+      handleRef.current === null
     }
+  }, [timeout])
 
+  const [returnValue] = useState(() => {
```

Comment:
Just curious, is `useState` better than `useRef` for this instance?

Reply:
this is instead of `useMemo`, it just ensures that that the value is always stable across the lifecycle of the component. A ref works too but i like that you can't change this unsafely vs a ref